In [8]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

## Q1. Pandas version
What's the version of Pandas that you installed?

In [9]:
print(f'Python Version is v{pd.__version__}') 

Python Version is v2.2.3


In [10]:
pd.__version__

'2.2.3'

In [11]:
url = r'https://raw.githubusercontent.com/alexeygrigorev/datasets/master/laptops.csv'

In [12]:
laptops = pd.read_csv(url)

In [13]:
laptops

Laptop       Status   Brand  \
0     ASUS ExpertBook B1 B1502CBA-EJ0436X Intel Core...          New    Asus   
1     Alurin Go Start Intel Celeron N4020/8GB/256GB ...          New  Alurin   
2     ASUS ExpertBook B1 B1502CBA-EJ0424X Intel Core...          New    Asus   
3     MSI Katana GF66 12UC-082XES Intel Core i7-1270...          New     MSI   
4     HP 15S-FQ5085NS Intel Core i5-1235U/16GB/512GB...          New      HP   
...                                                 ...          ...     ...   
2155  Razer Blade 17 FHD 360Hz Intel Core i7-11800H/...  Refurbished   Razer   
2156  Razer Blade 17 FHD 360Hz Intel Core i7-11800H/...  Refurbished   Razer   
2157  Razer Blade 17 FHD 360Hz Intel Core i7-11800H/...  Refurbished   Razer   
2158  Razer Book 13 Intel Evo Core i7-1165G7/16GB/1T...  Refurbished   Razer   
2159  Razer Book FHD+ Intel Evo Core i7-1165G7/16GB/...  Refurbished   Razer   

           Model                CPU  RAM  Storage Storage type       GPU  \
0     ExpertBook      Intel Core i5    8      512          SSD       NaN   
1             Go      Intel Celeron    8      256          SSD       NaN   
2     ExpertBook      Intel Core i3    8      256          SSD       NaN   
3         Katana      Intel Core i7   16     1000          SSD  RTX 3050   
4            15S      Intel Core i5   16      512          SSD       NaN   
...          ...                ...  ...      ...          ...       ...   
2155       Blade      Intel Core i7   16     1000          SSD  RTX 3060   
2156       Blade      Intel Core i7   16     1000          SSD  RTX 3070   
2157       Blade      Intel Core i7   32     1000          SSD  RTX 3080   
2158        Book  Intel Evo Core i7   16     1000          SSD       NaN   
2159        Book  Intel Evo Core i7   16      256          SSD       NaN   

      Screen Touch  Final Price  
0       15.6    No      1009.00  
1       15.6    No       299.00  
2       15.6    No       789.00  
3       15.6    No      1199.00  
4       15.6    No       669.01  
...      ...   ...          ...  
2155    17.3    No      2699.99  
2156    17.3    No      2899.99  
2157    17.3    No      3399.99  
2158    13.4   Yes      1899.99  
2159    13.4   Yes      1699.99  

[2160 rows x 12 columns]

## Q2. Records count
How many records are in the dataset?

In [14]:
records = laptops.shape[0]
print(f'Dataset records = {records} entries')

Dataset records = 2160 entries


In [15]:
records = laptops.shape[0]
print(f'Dataset records = {records} entries\nColumns = {laptops.shape[1]}')

Dataset records = 2160 entries
Columns = 12


## Q3. Laptop brands
How many laptop brands are presented in the dataset?

In [16]:
brands = pd.DataFrame(laptops['Brand'].unique())
brands = brands.value_counts().sum()
print(f'Laptop brands distinct count = {brands} brands')

Laptop brands distinct count = 27 brands


In [17]:
laptops['Brand'].nunique()

27

In [18]:
laptops.nunique()

Laptop          2160
Status             2
Brand             27
Model            121
CPU               28
RAM                9
Storage           12
Storage type       2
GPU               44
Screen            29
Touch              2
Final Price     1440
dtype: int64

## Q4. Missing values
How many columns in the dataset have missing values?

In [19]:
missing_values = laptops.isna().sum()
missing_values

Laptop             0
Status             0
Brand              0
Model              0
CPU                0
RAM                0
Storage            0
Storage type      42
GPU             1371
Screen             4
Touch              0
Final Price        0
dtype: int64

There are 3 columns with missing values in the dataset (Storage type - 42, GPU - 1371 and Screen - 4)

## Q5. Maximum final price
What's the maximum final price of Dell notebooks in the dataset?

In [20]:
# What’s the maximum final price of Dell notebooks in the dataset?

lap_final = pd.DataFrame(laptops[['Brand','Final Price']])
lap_dell = lap_final[lap_final['Brand']== 'Dell']
lap_dell['Final Price'].max()


np.float64(3936.0)

In [21]:
laptops.groupby('Brand')['Final Price'].agg(['min', 'mean', 'max'])

min         mean      max
Brand                                          
Acer               264.14  1001.285766  3691.00
Alurin             239.00   484.701379   869.00
Apple              299.00  1578.227672  3849.00
Asus               239.25  1269.380699  5758.14
Deep Gaming       1334.00  1505.377500  1639.01
Dell               379.00  1153.839881  3936.00
Denver             329.95   329.950000   329.95
Dynabook Toshiba   397.29   999.197895  1805.01
Gigabyte           799.00  1698.488958  3799.00
HP                 210.14   952.628478  5368.77
Innjoo             251.40   347.888333   431.38
Jetwing            469.27   469.270000   469.27
LG                 609.04  1730.028438  2399.00
Lenovo             239.00  1087.723880  5018.14
MSI                444.19  1913.451006  7150.47
Medion             246.89   905.331875  3799.00
Microsoft          599.01  1595.543247  3747.91
Millenium         1640.98  1976.845000  2312.71
PcCom             1049.00  1606.762917  1949.90
Primux             208.99   339.618750   599.41
Prixton            246.95   292.280000   329.95
Razer             1232.74  3265.701622  4999.01
Realme             999.00   999.000000   999.00
Samsung            365.90  1595.511364  3699.01
Thomson            201.05   316.860000   436.56
Toshiba            799.00   799.000000   799.00
Vant               632.00  1047.455455  1217.01

## Q6. Median value of Screen
Find the median value of Screen column in the dataset. <br />
Next, calculate the most frequent value of the same Screen column.<br />
Use fillna method to fill the missing values in Screen column with the most frequent value from the previous step.<br />
Now, calculate the median value of Screen once again.<br /> <br />
Has it changed?
<br />
Hint: refer to existing mode and median functions to complete the task.<br />
<br />
Yes<br />
No<br />

In [22]:
# 1. Find the median value of Screen column in the dataset. 

median_value = laptops['Screen'].median()
median_value

np.float64(15.6)

In [23]:
# 2. Next, calculate the most frequent value of the same Screen column.

most_frequent = laptops['Screen'].value_counts().idxmax()
most_frequent

np.float64(15.6)

In [24]:
# 3. Use fillna method to fill the missing values in Screen column with the most frequent value from the previous step. 
# laptops['Screen'].fillna(value=mean_screen_max, inplace=True)

laptops.fillna({'Screen' : most_frequent}, inplace=True)

In [25]:
# 4. Now, calculate the median value of Screen once again. Has it changed?
median_value = laptops['Screen'].median()
median_value

np.float64(15.6)

Has it changed?
<br />
No<br />

## Q7. Sum of weights
Select all the "Innjoo" laptops from the dataset.<br/><br/>
Select only columns RAM, Storage, Screen.<br/><br/>
Get the underlying NumPy array. Let's call it X.<br/><br/>
Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.<br/><br/>
Compute the inverse of XTX.<br/><br/>
Create an array y with values [1100, 1300, 800, 900, 1000, 1100].<br/><br/>
Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.<br/><br/>
What's the sum of all the elements of the result?<br/><br/>
Note: You just implemented linear regression. We'll talk about it in the next lesson.<br/><br/>
<br/>
0.43<br/>
45.29<br/>
45.58<br/>
91.30<br/>

In [27]:
laptops_injoo = laptops[laptops['Brand']=="Innjoo"]

In [28]:
laptops_injoo = laptops_injoo[['RAM', 'Storage', 'Screen']]

In [29]:
laptops_injoo

RAM  Storage  Screen
1478    8      256    15.6
1479    8      512    15.6
1480    4       64    14.1
1481    6       64    14.1
1482    6      128    14.1
1483    6      128    14.1

In [30]:
X = laptops_injoo.values
XTX = X.T.dot(X)

XTX_inv = np.linalg.inv(XTX)
XTX_inv

array([[ 2.78025381e-01, -1.51791334e-03, -1.00809855e-01],
       [-1.51791334e-03,  1.58286725e-05,  4.48052175e-04],
       [-1.00809855e-01,  4.48052175e-04,  3.87214888e-02]])

In [31]:
y = np.array([1100, 1300, 800, 900, 1000, 1100])

In [32]:
y

array([1100, 1300,  800,  900, 1000, 1100])

In [33]:
w = (XTX_inv @ X.T) @ y
w

array([45.58076606,  0.42783519, 45.29127938])

In [34]:
sum(w)

np.float64(91.2998806299555)

### Others

In [35]:
# What is the average RAM capacity (in GB) across all laptops in the dataset?
laptops["RAM"].mean()

np.float64(15.41388888888889)

In [36]:
laptops.describe().T

count         mean         std     min       25%       50%  \
RAM          2160.0    15.413889    9.867815    4.00    8.0000    16.000   
Storage      2160.0   596.294444  361.220506    0.00  256.0000   512.000   
Screen       2160.0    15.168912    1.202357   10.10   14.0000    15.600   
Final Price  2160.0  1312.638509  911.475417  201.05  661.0825  1031.945   

                 75%      max  
RAM            16.00   128.00  
Storage      1000.00  4000.00  
Screen         15.60    18.00  
Final Price  1708.97  7150.47

In [38]:
laptops.GPU.value_counts().sum()

np.int64(789)

In [39]:
# How many laptops in the dataset come with a dedicated GPU?

laptops.GPU.value_counts()

GPU
RTX 3050              129
RTX 3060              122
RTX 3070               97
RTX 4060               62
RTX 3080               51
GTX 1650               50
RTX 4070               40
RTX 4050               33
RTX 2070               25
RTX 2060               20
RTX 4080               18
RTX 4090               17
GTX 1660               14
RTX 2050               11
Radeon Pro 5500M        9
RTX A2000               8
MX 450                  7
RTX 2080                7
RTX A3000               7
T 1200                  6
RTX A1000               6
MX 330                  5
MX 550                  5
T 500                   5
T 550                   4
RX 6500M                3
GTX 1050                3
Radeon Pro RX 560X      3
RTX A5500               3
610 M                   2
RX 7600S                2
Radeon Pro 5300M        2
T 600                   2
A 730M                  1
A 370M                  1
RTX 3000                1
T 1000                  1
T 2000                  1
Radeon R

In [40]:
laptops.head(5)

Laptop Status   Brand  \
0  ASUS ExpertBook B1 B1502CBA-EJ0436X Intel Core...    New    Asus   
1  Alurin Go Start Intel Celeron N4020/8GB/256GB ...    New  Alurin   
2  ASUS ExpertBook B1 B1502CBA-EJ0424X Intel Core...    New    Asus   
3  MSI Katana GF66 12UC-082XES Intel Core i7-1270...    New     MSI   
4  HP 15S-FQ5085NS Intel Core i5-1235U/16GB/512GB...    New      HP   

        Model            CPU  RAM  Storage Storage type       GPU  Screen  \
0  ExpertBook  Intel Core i5    8      512          SSD       NaN    15.6   
1          Go  Intel Celeron    8      256          SSD       NaN    15.6   
2  ExpertBook  Intel Core i3    8      256          SSD       NaN    15.6   
3      Katana  Intel Core i7   16     1000          SSD  RTX 3050    15.6   
4         15S  Intel Core i5   16      512          SSD       NaN    15.6   

  Touch  Final Price  
0    No      1009.00  
1    No       299.00  
2    No       789.00  
3    No      1199.00  
4    No       669.01

In [41]:
# What is the final price of the cheapest laptop using an AMD processor in the dataset?
laptops.CPU.value_counts()

CPU
Intel Core i7            710
Intel Core i5            535
AMD Ryzen 7              156
Intel Core i3            130
AMD Ryzen 5              127
Intel Celeron             94
Intel Core i9             94
Intel Evo Core i7         82
AMD Ryzen 9               44
AMD Ryzen 3               44
Intel Evo Core i5         30
Apple M2                  28
AMD 3020e                 13
Apple M2 Pro              13
Apple M1                  11
AMD Athlon                10
Intel Pentium             10
Apple M1 Pro               7
Intel Core M3              5
Qualcomm Snapdragon 7      3
AMD 3015e                  3
Microsoft SQ1              3
AMD Radeon 9               2
Qualcomm Snapdragon 8      2
Intel Evo Core i9          1
AMD Radeon 5               1
AMD 3015Ce                 1
Mediatek MT8183            1
Name: count, dtype: int64

In [43]:
# What is the final price of the cheapest laptop using an AMD processor in the dataset?
fin_price = laptops.groupby('CPU')['Final Price'].min()
fin_price = pd.DataFrame(fin_price).sort_values(["Final Price"])
fin_price

Final Price
CPU                               
Intel Celeron               201.05
Intel Core i5               210.14
Intel Core i3               273.61
AMD 3020e                   299.00
Intel Pentium               309.91
AMD 3015Ce                  327.00
AMD Ryzen 3                 327.51
AMD Ryzen 5                 338.81
AMD Athlon                  349.00
AMD 3015e                   397.00
Intel Core i7               407.49
Qualcomm Snapdragon 8       436.56
Apple M1                    449.01
Mediatek MT8183             458.00
AMD Ryzen 7                 478.99
AMD Radeon 5                499.00
Qualcomm Snapdragon 7       531.00
Intel Core M3               549.00
Intel Evo Core i5           701.28
Intel Evo Core i7          1016.00
Microsoft SQ1              1049.00
AMD Ryzen 9                1242.39
Intel Core i9              1313.64
Apple M2                   1519.00
Apple M2 Pro               2209.00
Apple M1 Pro               2500.01
Intel Evo Core i9          3699.01
AMD Radeon 9               4753.32

In [44]:
# What is the final price of the cheapest laptop using an AMD processor in the dataset?
fin_price = laptops.groupby('CPU')['Final Price'].min()
fin_price = pd.DataFrame(fin_price).sort_values(["CPU"], ascending=True)
fin_price

Final Price
CPU                               
AMD 3015Ce                  327.00
AMD 3015e                   397.00
AMD 3020e                   299.00
AMD Athlon                  349.00
AMD Radeon 5                499.00
AMD Radeon 9               4753.32
AMD Ryzen 3                 327.51
AMD Ryzen 5                 338.81
AMD Ryzen 7                 478.99
AMD Ryzen 9                1242.39
Apple M1                    449.01
Apple M1 Pro               2500.01
Apple M2                   1519.00
Apple M2 Pro               2209.00
Intel Celeron               201.05
Intel Core M3               549.00
Intel Core i3               273.61
Intel Core i5               210.14
Intel Core i7               407.49
Intel Core i9              1313.64
Intel Evo Core i5           701.28
Intel Evo Core i7          1016.00
Intel Evo Core i9          3699.01
Intel Pentium               309.91
Mediatek MT8183             458.00
Microsoft SQ1              1049.00
Qualcomm Snapdragon 7       531.00
Qualcomm Snapdragon 8       436.56